android_app/app/src/main/java/com/health/openscale/core/bluetooth/BluetoothActiveEraBF06.java

```diff
@@ -0,0 +1,326 @@
+                        int height = (int) Math.ceil(selectedUser.getBodyHeight());
+                        int age = selectedUser.getAge();
+                        int gender = selectedUser.getGender() == Converters.Gender.FEMALE ? 0 : 1;
+
+                        calculateBIA(height, impedance, stableWeightKg, age, gender);
+                        // TODO: report results
+                    }
+
+                } else {
+                    Timber.w("Unsupported number of ADCs: %s", number);
+                }
+
+                stopMachineState();
+                break;
+
+            case 0xD7: // HR measured
+                int hr = pkt[0x03] & 0xff;
+                Timber.i("Measured heart rate: %d", hr);
+                resumeMachineState();
+
+                break;
+
+            case 0xD8: // historical measurement
+                parseHistoricalPacket(pkt);
+
+            default:
+                Timber.w("Unsupported packet [%d]: %s", packetType, byteInHex(pkt));
+        }
+
+    }
+
+    private byte[] withCorrectCS(byte[] pkt) {
+        byte[] fixed = Arrays.copyOf(pkt, pkt.length);
+        fixed[fixed.length - 1] = sumChecksum(fixed, 2, fixed.length - 3);
+        return fixed;
+    }
+
+    /**
+     * Calculate BIA parameters
+     * for now, using forumlas from
+     * <a href="https://isn.ucsd.edu/courses/beng186b/project/2021/Raj_Sunku_Tsujimoto_Measuring_body_composition_via_body_impedance.pdf">paper</a>
+     *
+     * TODO: replace with reverse-engineered library version
+     *
+     * @param heightCm
+     * @param impedanceOhm
+     * @param weightKg
+     * @param age - in years
+     * @param gender - 0 - female, 1 - male
+     */
+    private void calculateBIA(int heightCm, double impedanceOhm, float weightKg, int age, int gender) {
+        // FFM = 0.36(H2/Z) + 0.162H + 0.289W − 0.134A + 4.83G − 6.83
+        double fatFreeMass = (0.36d * (Math.pow(heightCm, 2) / impedanceOhm))
+                + (0.162d * heightCm)
+                + (0.289d * weightKg)
+                - (0.134 * age)
+                + (4.83 * gender)
+                - 6.83;
+
+        double fatMass = weightKg - fatFreeMass;
+        double bodyFat = fatMass / weightKg * 100.0;
+        Timber.i("FFM: %.2f, FM: %.2f, BF: %.1f%%", fatFreeMass, fatMass, bodyFat);
+    }
+
+    private void parseHistoricalPacket(byte[] pkt) {
+        Instant time = Instant.ofEpochSecond(Converters.fromUnsignedInt24Be(pkt, 3));
+        float weight = (Converters.fromUnsignedInt24Be(pkt, 0x08) & 0x03FFFF) / 1000.0f;
+        float weightLeft = Converters.fromUnsignedInt16Be(pkt, 0x0b) / 100.0f;
+        int hr = pkt[0x0d] & 0xff;
+        int adc = Converters.fromUnsignedInt16Be(pkt, 0x0f);
+        Timber.i("Historical measurement: %.3f kg, Weight Left: %.2f kg, HR: %d, ADC: %d", weight, weightLeft, hr, adc);
+        // TODO: store historical results
```

Comment:
maybe just a simple

```
ScaleMeasurement historyMeasurement = new ScaleMeasurement();
historyMeasurement.setWeight(weight);
addScaleMeasurement(historyMeasurement);
```

would do it?

Reply:
The problem is - I'm not the sole user of this scales ) So unless we have a proper "select and claim" mechanic, I would rather not claim everything automatically

Reply:
This should be solved with the settings->Bluetooth->Intelligence user assignment which automatically assign the user which have the nearest weight. You could also enable then ignore out of range measurements to ignore weights that out of the user weights.